R/utils.R

```diff
@@ -405,3 +405,83 @@ defunction <- function(x) {
+#'
+#' - **Pre-computed `.lockfile`**: User is able to provide his own pre-computed `.lockfile` by setting the path to the
+#' `.lockfile` through `options(teal.renv.lockfile = "")`. Then this function is not used.
+#' - **Working directory `.lockfile`**: If `options(teal.renv.lockfile)` is empty, by default `teal` will
+#' create an `implicit` type of the `.lockfile`, that uses `renv::dependencies()` to detect all R packages in the
+#' current project working directory.
+#' - **`DESCRIPTION` based `.lockfile`**: You can always include a `DESCRIPTION` file in your working directory and
+#' enable `.lockfile` creation based on this file. To do this, run `renv::settings$snapshot.type("explicit")`.
+#' Naming of `type` is the same as in `renv::snapshot()`. For the `"explicit"` type refer to
+#' `renv::settings$package.dependency.fields()`
+#' to see what `DESCRIPTION` fields are included in the `.lockfile`.
+#' - **Custom files based `.lockfile`**: If you want to specify custom files as a base for the `.lockfile`, then run
+#' `renv::settings$snapshot.type("custom")` and set `renv.snapshot.filter` option.
+#'
+#' @note
+#' This function computes the `.lockfile` as a `promises::future_promise` promise, while
+#' running the evaluation of the process on a separate worker. If `future::plan()` was set to something different than
+#' `future::sequential`, it reuses the parallel backed. If not, it sets `future::multisession` as a parallel plan with
+#' 2 workers. Then, `shiny::ExtendedTask()` is used to run asynchronous computations. If `teal` app started with
+#' `future::sequential` plan, is it set back once the task is finished.
+#'
+#' @section `.lockfile` usage:
+#' Once you have a `.lockfile` file, you can restore R project with `ren::init();renv::restore()`.
+#'
+#' @param close (`logical(1)`) Flag that indicates whether to set `future::plan` to `future::sequential` when finishing the task.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()], [renv::init()].
+#'
+#' @return (`character(1)`) the path to the `lockfile` created in a `tempdir()`.
+#'
+#' @keywords internal
+create_renv_lockfile <- function(close) {
+  checkmate::assert_flag(close)
+  promise <- promises::future_promise({
+    # Below is not a file in tempdir() directory.
+    # If a file is created in tempdir() it gets deleted on 'then(onFulfilled' part.
+    lockfile_path <- "teal_renv_lock.lock"
+    shiny::onStop(function() file.remove(lockfile_path))
+
+    renv_status <- utils::capture.output(
+      renv::snapshot(
+        lockfile = lockfile_path,
+        prompt = FALSE,
+        force = TRUE
+        # type = is taken from renv::settings$snapshot.type()
+      )
+    )
+    if (any(grepl("Lockfile written", renv_status))) {
+      logger::log_trace("lockfile created successfully.", namespace = "teal")
+    } else {
+      logger::log_trace("lockfile created with issues.", namespace = "teal")
+    }
```

Comment:
here you are dependent on the print statements which is not very robust IMHO
I think that you are looking for `utils::capture.output(res <- renv::snapshot(...))` and then condition on `res`

Reply:
> It's the `cat` that I grab with `utils::capture.output` that contains the information of whether the creation was successful or not.

That's exactly what I mean when saying that you are dependent on what `renv` puts to the console. This type of data is mean for a human reading and is very fragile I would say. It's better to rely on the raw data and not how it is printed out.

Can you derive the success criteria out of `res`? Or maybe try-block?

Reply:
We use `force = TRUE` so that the lockfile is always created, even though there might be issues with potential future recreation of the project, based on the lockfile. 

There are a couple of possible scenarios in here
- use `force = TRUE` and get a proper lockfile 
- use `force = TRUE` and get an improper lockfile based on which you may not be able to recreate the project
- use `force = FALSE` and get the app to crash as there was an error caused by renv

I would aim to create and share a lockfile, even though it might not be restorable. Once the lockfile is created, there is no information about the restorability or not and the only information of whether it was a proper creation or not was in what's printed to the console. Unfortunately.

What we can do, but I don't find it fancy, is to change the option to `force = FALSE`, and put it in a tryCatch, and if there was an error then you can write that it was impossible to create a lockfile. But I rather expose broken/unrestorable lockfile that has some information, than no lockfile at all.

Reply:
Thanks for a detailed explanation. I do support the decision you made. I was only wondering if `snapshot()` returns the status somehow. If not then I guess we need to live with that. Then what you have done is correct. I would only rename `renv_status` to maybe `renv_logs` but it's just a small detail

Reply:
Coolio. Just renamed `renv_status` to `renv_logs`. Closing this thread 570b559e3c3f81a9f00a8b3c2510d5828784d134